### ISSUE TITLE
[new] Check if variable type was changed

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#90]
Happy new year! Let's starts 2016... I say... 2017 with new suggestions!

This is about to check if variable type was changed for some new assignment. We can think about two cases: _strong type changing checking_ and _weak type changing check_.

**What is a type change**: basically, if you declare a variable with some value, it should receive a type hint (except for `null`). For instance:

```
<?php

$undefined = null;            // Starts without a type hint
$number = 123;                // Starts as number
$string = '123';              // Starts as string
$bool = true;                 // Starts as bool
$array = [];                  // Starts as array
$function = function () { };  // Starts as function (so on)
$user = new User();           // Starts as User (that extends People)
```

Then, if I change a `$undefined` to a new value, it will be consider this next assignment. Then:

```
<?php

$undefined = null;    // Starts without a type hint
$undefined = 123;     // Now it is a number|null
```

For other types, we should consider the variable type changing checking. Then I think about two ways:

**Strong type changing checking**: it should consider any type changing (except for first `null`), even for scalar. Then now we should have:

```
<?php

$number = 123;          // Starts as number
$number = 456;          // Valid: keeped as number
$number = '789';        // Invalid: changed to string
$number = new User();   // Invalid: changed to User
$number = [];           // Invalid: changed to array

$undefined = null;      // Starts without a type hint
$undefined = 123;       // Valid: set as number
$undefined = null;      // Invalid: changed to null
$undefined = '123';     // Invalid: changed to string

$user = new User();     // Starts as User (that extends People)
$user = new ProUser();  // Valid: using as ProUser (that extends User)
$user = new People();   // Invalid: changed to People (should keep to be User)

$array = [];            // Starts as array
$array = [1, 2, 3];     // Valid: keeped as array
$array[] = 4;           // Valid: keeped as array
$array = '123';         // Invalid: changed to number
```

**Weak type chaning checking**: is a low strict version of above. It should allows you changes any scalar to any scalar, including null (only if it starts with `null`). Now see what changes (the `***` denotes changes):

```
<?php

$number = 123;          // Starts as number
$number = 456;          // Valid: keeped as number
$number = '789';        // Valid: changed to number|string ***
$number = new User();   // Invalid: changed to User
$number = [];           // Invalid: changed to array

$undefined = null;      // Starts without a type hint
$undefined = 123;       // Valid: set as number|null ***
$undefined = null;      // Valid: keeped as number|null ***
$undefined = '123';     // Invalid: changed to string|number|null ***

$user = new User();     // Starts as User (that extends People)
$user = new ProUser();  // Valid: using as ProUser (that extends User)
$user = new People();   // Valid: changed to User|People ***
$array = [];            // Starts as array
$array = [1, 2, 3];     // Valid: keeped as array
$array[] = 4;           // Valid: keeped as array
$array = '123';         // Invalid: changed to number
```

**Annotations should be strictify variables**: if you declare the type using Annotation, it should be treated always strictly (strongly) - it'll be great if autofix is available here. Then, for instance:

```
<?php

/** @var null $undefined */
$undefined = null;          // Valid: keeped as null
$undefined = 123;           // Invalid: changed to number
$undefined = null;          // Valid: back to null, that is valid

/** @var number|null $number */
$number = 123;              // Valid: starts as number
$number = null;             // Valid: changed to null
$number = '123';            // Invalid: changed to string
```

Uff! Is that.

**Comment by funivan** [comment]
```php
$number = 123;          // Starts as number
$number = '789';        // Valid: changed to number|string ***
```
I think number is invalid. Why we should keep this variable valid? 


And here is another case:
```php
$undefined = null;      // Starts without a type hint
$undefined = 123;       // Valid: set as number
$undefined = null;      // Invalid: changed to null
$undefined = '123';     // Invalid: changed to string
```

If we define `$undefined=123` and then we want to clean this value we can use `$undefined=null` so all is ok. IMHO

**Comment by rentalhost** [comment]
@funivan Your **first example** will be truth only on the _strong checking_: you can't change the type after the first declaration (firstly you have declared a number to variable, then you redeclared it as string, then it should be invalid). For weak checking you can do that if it is a strict scalar or null (int|float|string|null) - it should excludes boolean, I guess.

Your **second exemple** too is only for _strong checking_: it'll allow that you initialize a variable with `null` but, after you set a "final type" you can't change anymore (even for _null_ again). For weak checking you can do that.

This second example is interesting (on _strong checking_) because force you to not uninitialize the variable after you start it (but you can set it to _zero_). It'll assure that your variable will be ever a number after initialized:

```php
$number = null;
echo $number; // Warning: variable uninitialized.

$number = 123;
echo $number; // Valid.

$number = 0;
echo $number; // Valid.

$number = null; // Warning: variable deinitialized.
echo $number; // Warning: variable uninitialized.
```

In this last case, you still can annotate this variable to accept null, even by using _strong checking_, for instance: `@var number|null $number`.

**Comment by kalessil** [comment]
Some clarifications needed (frankly I'm in doubt if this applicable to PHP and beneficial in current project state):

- PHP APIs are often using `false` as the `null` in use-cases, did you notice that?
- What about `mixed` types?

**Comment by rentalhost** [comment]
@kalessil I don't understand your first line about PHP APIs, can you tell some example to I think about that? About `mixed` it should be accept anything (basically, it could be ignored on analysis because will acts like it works today).

**Comment by kalessil** [comment]
@rentalhost: e.g. http://php.net/manual/en/function.substr.php returns false in case of failures.

**Comment by rentalhost** [comment]
Right. It should not be a problem because in the major cases that you should use it is on `if()`, for instance. But if you use it on some variable, you should annotate on it. Example:

```php
/** @var false|string $substr */
$substr = substr($mystring, 0, 5);

/** @var mixed $substr */
$substr = substr($mystring, 0, 5);

// Then...
$substr = 123; // Invalid
$substr = false; // Valid
$substr = '123'; // Valid
```

The PHPStorm has annotated the `substr()` with incorrect `@return` (it only consider strings, but from doc is `string|false`), but it is an issue that should be fixed by they.
